examples/benchmark.py

```diff
@@ -31,183 +124,469 @@ def downloadAmberSuite():
+
+# Attempt to get GPU info
+try:
+    import subprocess
+    cmd = 'nvidia-smi --query-gpu=driver_version,gpu_name --format=csv,noheader'
+    output = subprocess.check_output(cmd, shell=True, text=True)
+    system_info['nvidia_driver'], system_info['gpu'] = output.strip().split(', ')
+except Exception as e:
+    pass
+
+for key, value in system_info.items():
+    print(f'{key}: {value}')
+
+if args.outfile is not None:
+    # Remove output file
+    import os
+    if os.path.exists(args.outfile):
+        os.unlink(args.outfile)
+    # Write system info
+    appendTestResult(args.outfile, system_info=system_info)
+
+if args.platform is not None:
+    # Use specified subset of platforms
+    requested_platforms = args.platform.split(',')
+    if not set(requested_platforms).issubset(PLATFORMS):
+        parser.error(f'Available platforms: {PLATFORMS}')
+    PLATFORMS = requested_platforms
+
+if args.test is not None:
+    # Use specified subset of tests
+    requested_tests = args.test.split(',')
+    if not set(requested_tests).issubset(TESTS):
+        parser.error(f'Available tests: {TESTS}')
+    TESTS = requested_tests
+
+if args.precision is not None:
+    # Use specified subset of precisions
+    requested_precisions = args.precision.split(',')
+    if not set(requested_precisions).issubset(PRECISIONS):
+        parser.error(f'Available precisions: {PRECISIONS}')
+    PRECISIONS = requested_precisions
+else:
+    PRECISIONS = ['single'] # Peter's preferred default
+
+if args.ensemble is not None:
+    # Use specified subset of ensembles
+    requested_ensembles = args.ensemble.split(',')
+    if not set(requested_ensembles).issubset(ENSEMBLES):
+        parser.error(f'Available ensembles: {ENSEMBLES}')
+    ENSEMBLES = requested_ensembles
+else:
+    ENSEMBLES = ['NVT'] # Peter's preferred default
+
+if args.bond_constraints is not None:
+    # Use specified subset of constraints
+    requested_bond_constraints = args.bond_constraints.split(',')
+    if not set(requested_bond_constraints).issubset(BOND_CONSTRAINTS):
+        parser.error(f'Available bond constraints: {BOND_CONSTRAINTS}')
+    BOND_CONSTRAINTS = requested_bond_constraints
+else:
+    BOND_CONSTRAINTS = ['hbonds'] # Peter's preferred default
+
+# Combinatorially run all requested benchmarks, ignoring combinations that cannot be run
+from openmm import OpenMMException
+from itertools import product
+
+if args.style == 'simple':
+    for (test, args.bond_constraints, args.ensemble, args.platform, args.precision) in product(TESTS, BOND_CONSTRAINTS, ENSEMBLES, PLATFORMS, PRECISIONS):
         try:
             runOneTest(test, args)
-        except Exception as ex:
-            print('Test failed: %s' % ex.message)
+        except OpenMMException as e:
+            if args.verbose:
+                print(e)
+            pass
+elif args.style == 'rich':
+    from rich.live import Live
+    from rich.table import Table
+
+    from rich import box
```

Comment:
I'm not familiar with `rich`.  We don't document anywhere that it's a dependency.  And adding a new dependency just for this seems excessive.

Reply:
[`rich`](https://github.com/Textualize/rich) is an amazingly popular (1.2M downloads) flexible terminal formatting library.

It's not a dependency---it's optional and not enabled by default. The user needs to specify `--style rich` to enable this.

Reply:
An optional dependency is still a dependency.  It's something that needs to be tested, and that will break our code when they make changes to future versions.

Let's use a simple tab or comma delimited output.  It may not look quite as pretty on the screen, but that's not a serious concern for benchmarking.  And it will be very easy to copy and paste into a spreadsheet.

Reply:
> An optional dependency is still a dependency. 

That makes absolutely no sense.

> It's something that needs to be tested, and that will break our code when they make changes to future versions.

We are doing zero continuous integration testing right now, so this is irrelevant.

> Let's use a simple tab or comma delimited output. It may not look quite as pretty on the screen, but that's not a serious concern for benchmarking. And it will be very easy to copy and paste into a spreadsheet.

Happy for you to implement this if you want, but this library is orders of magnitude more popular than OpenMM, so I don't agree there is a serious issue with this being a problematic optional dependency.

Reply:
> We are doing zero continuous integration testing right now, so this is irrelevant.

All code needs to be tested, whether it's done through CI or manually.  Do you think the benchmarking script has never been tested?  And if it's not being tested through CI, that's all the more recent not to introduce dependencies that could cause it to spontaneously break at unknown times.

> Happy for you to implement this if you want

Ok, I'll make the changes and push them to your branch.

Reply:
> All code needs to be tested, whether it's done through CI or manually. Do you think the benchmarking script has never been tested? And if it's not being tested through CI, that's all the more recent not to introduce dependencies that could cause it to spontaneously break at unknown times.

No, I'm just saying that an equivalent level of testing is being performed now as was performed before. Nothing has degraded by introducing this optional library---I have tested it with and without.

> We are doing zero continuous integration testing right now, so this is irrelevant.

All code needs to be tested, whether it's done through CI or manually. Do you think the benchmarking script has never been tested? And if it's not being tested through CI, that's all the more recent not to introduce dependencies that could cause it to spontaneously break at unknown times.

> Ok, I'll make the changes and push them to your branch.

Feel free to. I've run out of time to implement things you clearly want. The only thing I care about in this PR is the ability for others to run the benchmarking script and generate a human- and machine-readable artifact they can share with us. It would be even better if we could automate the deposition of that artifact somewhere central, but I don't have a good idea about how to easily do this yet.